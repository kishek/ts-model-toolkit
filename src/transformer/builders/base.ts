import { SourceFile, SyntaxKind } from 'ts-morph';

import { Transformer } from '..';
import { ParserResult } from '../../parser/types';

export class BuilderBaseTransformer extends Transformer {
  /**
   * Translates between a parser type and a TypeScript type.
   */
  protected getPropertyType(
    type: ParserResult.Type,
    imports: ParserResult.Import[],
    builderFile: SourceFile,
    typeParameters: ParserResult.Type[] = [],
  ): string {
    // Small wrapper for returning some types with imports.
    const withImport = (rawType = type) => {
      this.addImport({
        importIdentifier: rawType.raw,
        imports,
        file: builderFile,
      });
      return rawType.raw;
    };

    // NOTE: if a type has been supplied, this means that the
    // current type needs to be overriden.
    if (type.suppliedValue) {
      return this.getPropertyType(
        type.suppliedValue,
        imports,
        builderFile,
        typeParameters,
      );
    }

    switch (type.kind.value) {
      case SyntaxKind.StringKeyword:
        return 'string';
      case SyntaxKind.NumberKeyword:
        return 'number';
      case SyntaxKind.BooleanKeyword:
        return 'boolean';
      case SyntaxKind.InterfaceDeclaration:
        return type.raw;
      case SyntaxKind.Identifier:
        if (type.defaultValue) {
          return `${withImport(type)}.${type.defaultValue.raw}`;
        }
        return type.raw;
      case SyntaxKind.TypeReference: {
        if (type.defaultValue || type.constraintValue) {
          // NOTE: in this case, we make a strong assumption that the type parameter
          // required by this property is already a type parameter on the interface.
          // In this case, the builder statement can fallback to the 'generic' type
          // parameter which has been defined as part of the interface.
          // --------------------------------------------------------------------
          // TODO: add support for 'finding' the correct type parameter.
          if (typeParameters.length === 1) {
            return withImport(type);
          }
          return withImport(type.defaultValue || type.constraintValue);
        }
        if (type.typeArguments?.length === 1) {
          const typeArgument = withImport(type.typeArguments[0]);
          const typeReference = withImport(type);
          return `${typeReference}<${typeArgument}>`;
        }
        return withImport(type);
      }
      case SyntaxKind.ArrayType: {
        const elements = type.elements;
        if (elements) {
          const [element] = elements;
          return `${this.getPropertyType(
            element,
            imports,
            builderFile,
            typeParameters,
          )}[]`;
        }
      }
      case SyntaxKind.UnionType: {
        const elements = type.elements;
        if (elements) {
          return elements
            .map((element) =>
              this.getPropertyType(element, imports, builderFile, typeParameters),
            )
            .join(' | ');
        }
      }
      case SyntaxKind.TupleType: {
        const elements = type.elements;
        if (elements) {
          const [maybeRequired, maybeRest] = elements;
          // NOTE: we relax tuple types into flat arrays if and only if they
          // are composed of a rest type. The rest type will inherently
          // capture the narrowing type at the start, however due to the
          // limitations of GraphQL, cannot be enforced in the schema itself.
          if (maybeRest && maybeRest.kind.value === SyntaxKind.RestType) {
            if (maybeRest.elements) {
              const restElement = maybeRest.elements[0];
              const requiredType = this.getPropertyType(
                maybeRequired,
                imports,
                builderFile,
                typeParameters,
              );
              const restType = this.getPropertyType(
                restElement,
                imports,
                builderFile,
                typeParameters,
              );
              return `[${requiredType}, ...${restType}]`;
            }
          }
        }
      }
      default:
        console.warn(`Woah! ${type.kind.name} unsupported by builder transformer ðŸ˜±`);
        return 'Unknown';
    }
  }
}
