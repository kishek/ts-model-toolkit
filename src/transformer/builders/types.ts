import { SourceFileTransformerOpts } from '../types';

export interface BuilderTransformerOpts extends SourceFileTransformerOpts {}
