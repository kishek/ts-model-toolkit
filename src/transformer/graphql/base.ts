import { SyntaxKind } from 'ts-morph';

import { GraphQLTransformerOpts } from './types';
import { Transformer } from '../';
import { ParserResult } from '../../parser/types';

/**
 * Special mappings specific to GraphQL.
 */
export const GQL_MAPPINGS: Record<string, string> = {
  Date: 'String',
};

export class GraphQLBaseTransformer extends Transformer {
  /**
   * A method which translates parser type representations to GQL.
   * @param type
   */
  protected getGraphQLType(type: ParserResult.Type, tags?: [string, string][]): string {
    const gqlType = this.getGraphQLFromType(type, tags);
    if (GQL_MAPPINGS[gqlType]) {
      return GQL_MAPPINGS[gqlType];
    }
    return gqlType;
  }

  private getGraphQLFromType(type: ParserResult.Type, tags?: [string, string][]): string {
    // NOTE: if a type has been supplied, this means that the
    // current type needs to be overriden.
    if (type.suppliedValue) {
      return this.getGraphQLFromType(type.suppliedValue, tags);
    }

    switch (type.kind.value) {
      case SyntaxKind.StringKeyword:
        return 'String';
      case SyntaxKind.NumberKeyword:
        console.log('?', tags);
        if (tags?.some((t) => t[0] === 'float')) {
          return 'Float';
        }

        return 'Int';
      case SyntaxKind.BooleanKeyword:
        return 'Boolean';
      case SyntaxKind.EnumMember:
        return type.defaultValue?.raw || type.raw;
      case SyntaxKind.InterfaceDeclaration:
      case SyntaxKind.Identifier:
        return type.raw;
      case SyntaxKind.TypeReference: {
        if (type.defaultValue) {
          return type.defaultValue.raw;
        }
        if (type.constraintValue) {
          return type.constraintValue.raw;
        }
        if (type.typeArguments && type.typeArguments.length === 1) {
          return type.typeArguments[0].raw;
        }
        return type.raw;
      }
      case SyntaxKind.ArrayType: {
        const elements = type.elements;
        if (elements) {
          const [element] = elements;
          return `[${this.getGraphQLFromType(element)}]`;
        }
      }
      case SyntaxKind.UnionType: {
        const elements = type.elements;
        if (elements) {
          return elements.map((e) => this.getGraphQLFromType(e, tags)).join(' | ');
        }
      }
      case SyntaxKind.TupleType: {
        const elements = type.elements;
        if (elements) {
          const [, maybeRest] = elements;
          // NOTE: we relax tuple types into flat arrays if and only if they
          // are composed of a rest type. The rest type will inherently
          // capture the narrowing type at the start, however due to the
          // limitations of GraphQL, cannot be enforced in the schema itself.
          if (maybeRest && maybeRest.kind.value === SyntaxKind.RestType) {
            if (maybeRest.elements) {
              return this.getGraphQLFromType(maybeRest.elements[0]);
            }
          }
        }
      }
      default:
        console.warn(`Woah! ${type.kind.name} unsupported by GraphQL transformer ðŸ˜±`);
        return 'Unknown';
    }
  }

  public getGraphQLNullableMarker(
    property: ParserResult.Property,
    opts?: GraphQLTransformerOpts,
  ): string {
    if (opts?.inheritNullabilityFromStructure) {
      if (property.isRequired) {
        return '!';
      }
      return '';
    }
    return '';
  }
}
