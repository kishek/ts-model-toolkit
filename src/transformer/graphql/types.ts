export interface GraphQLTransformerOpts extends Record<string, unknown> {
  /**
   * Indicates whether the 'required' or 'not required' marker in TypeScript
   * needs to be reflected back into the GraphQL representation.
   */
  inheritNullabilityFromStructure?: boolean;
  /**
   * Indicates whether the `@key` directive needs to be attached to generated
   * GQL files, to support Apollo Federation.
   * @link https://www.apollographql.com/docs/federation/#federated-schema-example
   */
  useApolloFederationKeyDirective?: boolean;
  /**
   * Indicates whether the `@shareable` directive needs to be attached to generated
   * GQL files, to support Apollo Federation.
   * @link https://www.apollographql.com/docs/federation/federated-types/federated-directives/#shareable
   */
  useApolloFederationShareableDirective?: boolean;
  /**
   * Indicates whether the Relay connection specification should be used for certain
   * qualifiers. These are supplied as parameters below.
   * @link https://relay.dev/graphql/connections.htm
   */
  useRelayConnectionSpecification?: {
    /**
     * The qualifiers for which the Relay connection specification should be used.
     */
    qualifiers: string[];
  };
  /**
   * Indicates whether the current structure should be treated as a GraphQL input type.
   */
  inputType?: {
    /**
     * The resolver type for which an input type is being generated.
     */
    type: 'query' | 'mutation';
    /**
     * Transformer for the GraphQL input schema's name.
     */
    inputNameTransformer: (name: string) => string;
    /**
     * Transformer for the GraphQL resolver's name.
     */
    resolverNameTransformer?: (name: string) => string;
  };
}

export interface GraphQLTransformPropertyResult {
  /**
   * The raw GraphQL result of the transformer.
   */
  graphql: string;
  /**
   * The additional GraphQL declarations which the `graphql` result depends on.
   */
  additionalDeclarations: string[];
}
