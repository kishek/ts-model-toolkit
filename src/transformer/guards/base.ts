import { FunctionDeclaration, SourceFile, SyntaxKind } from 'ts-morph';

import { GuardTransformerOpts } from './types';
import { Transformer } from '../';
import { ParserResult } from '../../parser/types';

export interface GetGuardOpts {
  type: ParserResult.Type;
  inputVariable?: string;
  imports: ParserResult.Import[];
  file: SourceFile;
  fileOriginal: SourceFile;
}

/**
 * Special checks done for the guard transformer only.
 */
export const GUARD_MAPPINGS: Record<string, (variable: string) => string> = {
  Date: (variable) => `${variable}.constructor.name === "Date"`,
  URL: (variable) => `${variable}.constructor.name === "URL"`,
};

export class GuardBaseTransformer extends Transformer {
  /**
   * A method which translates parser type representations to guard checks.
   * @param type
   */
  protected getGuard({
    type,
    inputVariable = 'entity',
    imports,
    file,
    fileOriginal,
  }: GetGuardOpts): string {
    const defaultOpts = { inputVariable, imports, file, fileOriginal };
    // All references to constituting type guards are handled here.
    const toGuardWithImport = (rawType = type) => {
      if (GUARD_MAPPINGS[rawType.raw]) {
        return GUARD_MAPPINGS[rawType.raw](inputVariable);
      }
      const fnName = `is${rawType.raw}`;
      // Add import to where the sub-guard comes from.
      this.addImport({
        importIdentifier: rawType.raw,
        importName: fnName,
        imports,
        file,
        importFormatter: this.formatImport,
        useEntrypoint: true,
      });
      return `${fnName}(${inputVariable})`;
    };

    // NOTE: if a type has been supplied, this means that the
    // current type needs to be overriden.
    if (type.suppliedValue) {
      return this.getGuard({ ...defaultOpts, type: type.suppliedValue });
    }

    switch (type.kind.value) {
      case SyntaxKind.StringKeyword:
        return `typeof ${inputVariable} === "string"`;
      case SyntaxKind.NumberKeyword:
        return `typeof ${inputVariable} === "number"`;
      case SyntaxKind.BooleanKeyword:
        return `typeof ${inputVariable} === "boolean"`;
      case SyntaxKind.Identifier: {
        if (type.defaultValue) {
          this.addImport({
            importIdentifier: type.raw,
            imports,
            file: fileOriginal,
            creatingFile: file,
            useDirectory: true,
          });
          return `${inputVariable} === ${type.raw}.${type.defaultValue.raw}`;
        }
      }
      case SyntaxKind.InterfaceDeclaration:
      case SyntaxKind.TypeReference: {
        const fallbackType = type.defaultValue || type.constraintValue;
        if (fallbackType) {
          const fnName = `is${type.raw}`;
          const fnNameFallback = `is${fallbackType.raw}`;

          this.addImport({
            importIdentifier: fallbackType.raw,
            importName: fnNameFallback,
            imports,
            file,
            importFormatter: this.formatImport,
            useEntrypoint: true,
          });

          return `(${fnName} || ${fnNameFallback})(${inputVariable})`;
        }
        if (type.typeArguments?.length === 1) {
          const typeArgument = type.typeArguments[0];
          const typeArgumentFnName = `is${typeArgument.raw}`;

          // Add imports for type argument and its associated guard.
          this.addImport({
            importIdentifier: typeArgument.raw,
            imports,
            file: fileOriginal,
            creatingFile: file,
            useDirectory: true,
          });
          this.addImport({
            importIdentifier: typeArgument.raw,
            importName: typeArgumentFnName,
            imports,
            file,
            importFormatter: this.formatImport,
            useEntrypoint: true,
          });
          // Add import for type guard and type guard for type parameter.
          toGuardWithImport(type);
          toGuardWithImport(typeArgument);
          // Construct a generic-aware guard statement.
          return `is${type.raw}<${typeArgument.raw}>(${inputVariable}, ${typeArgumentFnName})`;
        }

        return toGuardWithImport();
      }
      case SyntaxKind.ArrayType: {
        const elements = type.elements;
        if (elements) {
          const [element] = elements;
          return `Array.isArray(${inputVariable}) && ${inputVariable}.every((item: any) => ${this.getGuard(
            {
              ...defaultOpts,
              type: element,
              inputVariable: 'item',
            },
          )})`;
        }
      }
      case SyntaxKind.UnionType: {
        const elements = type.elements;
        if (elements) {
          return `(${elements
            .map((element) => this.getGuard({ ...defaultOpts, type: element }))
            .join(' || ')})`;
        }
      }
      case SyntaxKind.TupleType: {
        const elements = type.elements;
        if (elements && elements.length > 0) {
          const [maybeRequired, maybeRest] = elements;
          // NOTE: we enforce the guards for our tuple types based on what is 'required'
          // as an element of the tuple, and what the rest of the elements need to be.
          if (maybeRest && maybeRest.kind.value === SyntaxKind.RestType) {
            if (maybeRest.elements && maybeRest.elements.length > 0) {
              const restType = maybeRest.elements[0];
              const restGuard = this.getGuard({ ...defaultOpts, type: restType });
              const requiredGuard = this.getGuard({
                ...defaultOpts,
                type: maybeRequired,
                inputVariable: 'item',
              });
              return `Array.isArray(${inputVariable}) && ${restGuard} && ${inputVariable}.some(item => ${requiredGuard})`;
            }
          }
        }
      }
      default:
        console.warn(`Woah! ${type.kind.name} unsupported by Guard transformer ðŸ˜±`);
        return 'false';
    }
  }
  /**
   * A method which returns a reference to a template file (for the purposes of generating type guard)
   * and the function in which the guard will live.
   */
  protected getTypeGuardReferences(
    structure: ParserResult.Structure,
    opts: GuardTransformerOpts,
    originalFileRef: SourceFile,
  ) {
    const fileRef = this.getFileRef(opts);
    const imports = this.getAllImports(structure);

    const typeParameters = this.getTypeParameters(
      structure,
      imports,
      fileRef,
      originalFileRef,
    );
    const typeParametersInReturn = typeParameters.map((t) => t.name).join(', ');

    const functionRef = fileRef.addFunction({
      name: `is${structure.name}`,
      isExported: true,
      typeParameters,
      returnType: `entity is ${structure.name}${
        typeParametersInReturn ? `<${typeParametersInReturn}>` : ''
      }`,
      parameters: [
        {
          name: 'entity',
          type: 'any',
        },
      ],
    });

    // NOTE: Add a parameter for a guard which is passed to 'generic' guards. This is to ensure that
    // we check the top-level type when extending a generic type, rather than the base-level, as we
    // may miss a whole host of properties when performing this check.
    // ------------------------------------------------
    // We only support this for cases where there is *one* type parameter at the moment.
    if (typeParameters.length === 1) {
      const typeParameterName = typeParameters[0].name;
      functionRef.addParameter({
        name: `is${typeParameterName}`,
        type: `(entity: any) => entity is ${typeParameterName}`,
        hasQuestionToken: true,
        leadingTrivia: '// eslint-disable-next-line @typescript-eslint/no-unused-vars',
      });
    }
    if (typeParameters.length > 1) {
      console.warn(
        `Oh no! Received ${typeParameters.length} type parameters for ${structure.name} ðŸ™€ This is not supported by the 'guards' transformer in \`ts-model-toolkit\` at the moment.`,
      );
    }

    return {
      fileRef,
      functionRef,
      imports,
    };
  }

  /**
   * A method which places the predicates for a type guard into a well-structured code block
   * inside of the function reference.
   */
  protected addFunctionBody(functionRef: FunctionDeclaration, guard: string): void {
    functionRef.setBodyText((writer) => {
      const guardStatement = guard.length > 0 ? `&& (${guard})` : '';
      writer.writeLine(`if (entity ${guardStatement})`).block(() => {
        writer.writeLine('return true;');
      });
      writer.writeLine('return false;');
    });
  }

  /**
   * An import formatter used by default by the guard transformer.
   */
  protected formatImport(importDef: ParserResult.Import): string {
    if (importDef.path.startsWith('@')) {
      return importDef.path + '/guards';
    }

    return importDef.path + '.guard';
  }
}
