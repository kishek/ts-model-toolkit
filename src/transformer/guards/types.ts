import { SourceFileTransformerOpts } from '../types';

export interface GuardTransformerOpts extends SourceFileTransformerOpts {}
