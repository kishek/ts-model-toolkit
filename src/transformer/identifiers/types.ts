import { SourceFileTransformerOpts } from '../types';

interface IdentifierTransformerImportOpts {
  /**
   * Name of the declaration to import.
   */
  name: string;
  /**
   * Path to the import (absolute).
   */
  path: string;
}

export interface IdentifierTransformerOpts extends SourceFileTransformerOpts {
  /**
   * The name of the structure which a structure needs to extend to be
   * reference-able using an ID.
   */
  mustExtendStructureOfName: string;
  /**
   * The import declarations for the base identifier class to be extended, and
   * the parser which exposes a `.parse` method.
   */
  imports: {
    /**
     * Location for the base class.
     */
    base: IdentifierTransformerImportOpts;
    /**
     * Location for the parser class.
     */
    parser: IdentifierTransformerImportOpts;
  };
}
