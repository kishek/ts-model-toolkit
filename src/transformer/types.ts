import { Project, SourceFile } from 'ts-morph';

import { ParserResult } from '../parser/types';

export interface AddImportOpts {
  importIdentifier: string;
  importName?: string;
  imports: ParserResult.Import[];
  file: SourceFile;
  creatingFile?: SourceFile;
  importFormatter?: (importDef: ParserResult.Import) => string;
  useDirectory?: boolean;
  useEntrypoint?: boolean;
  debug?: boolean;
}

export interface SourceFileTransformerOpts extends Record<string, unknown> {
  /**
   * The project in which the model file lives. This is needed in order to secure imports
   * correctly for the file.
   */
  project: Project;
  /**
   * The model file for which a guard will be created.
   */
  modelFile: SourceFile;
  /**
   * The output path for the type guard. This is where the file will be written to.
   */
  outputPath: string;
}
