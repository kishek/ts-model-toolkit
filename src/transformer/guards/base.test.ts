import { Project } from 'ts-morph';

import { GetGuardOpts, GuardBaseTransformer } from './base';
import { ParserResult } from '../../parser/types';
import { testBooleanType } from '../__mocks__/test.type.boolean';
import { testIdentifierType } from '../__mocks__/test.type.identifier';
import { testInterfaceType } from '../__mocks__/test.type.interface';
import { testNumberType } from '../__mocks__/test.type.number';
import { testStringType } from '../__mocks__/test.type.string';
import { testTupleAndRestType } from '../__mocks__/test.type.tuple-and-rest-types';
import { testTypeAliasType } from '../__mocks__/test.type.type-alias';
import { testTypeArgumentType } from '../__mocks__/test.type.type-argument';
import { testTypeReferenceType } from '../__mocks__/test.type.type-reference-with-array';
import { testTypeReferenceGenericWithDefaultAndConstraintType } from '../__mocks__/test.type.type-reference-with-generic';
import { testUnionType } from '../__mocks__/test.type.union-type';
import { testUnknownType } from '../__mocks__/test.type.unknown';

// Below taken from assorted snapshots.
const testTypes: ParserResult.Type[] = [
  testStringType,
  testNumberType,
  testBooleanType,
  testIdentifierType,
  testInterfaceType,
  testTypeReferenceType,
  testTypeReferenceGenericWithDefaultAndConstraintType,
  testTypeAliasType,
  testUnionType,
  testTupleAndRestType,
  testTypeArgumentType,
];

class TestGraphQLTransformer extends GuardBaseTransformer {
  public getGuardProxy(opts: GetGuardOpts): string {
    return this.getGuard(opts);
  }
}

beforeEach(() => {
  jest.spyOn(console, 'warn').mockImplementation();
});

afterEach(jest.clearAllMocks);

describe('GuardBaseTransformer', () => {
  testTypes.forEach((testType) => {
    it(`constructs correct type guard for Type.${testType.kind.name}`, () => {
      const transformer = new TestGraphQLTransformer();
      expect(
        transformer.getGuardProxy({
          type: testType,
          imports: [],
          file: new Project().createSourceFile('foo.ts'),
        }),
      ).toMatchSnapshot();
    });
  });

  it('falls back to `false` to invalidate type guard when unsupported kind supplied', () => {
    const transformer = new TestGraphQLTransformer();
    expect(
      transformer.getGuardProxy({
        type: testUnknownType,
        imports: [],
        file: new Project().createSourceFile('foo.ts'),
      }),
    ).toMatchInlineSnapshot(`"false"`);
  });
});
