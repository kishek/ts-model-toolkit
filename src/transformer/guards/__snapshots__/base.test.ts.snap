// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GuardBaseTransformer constructs correct type guard for Type.ArrayType 1`] = `"Array.isArray(entity) && entity.every((item: any) => isTestEnum(item))"`;

exports[`GuardBaseTransformer constructs correct type guard for Type.BooleanKeyword 1`] = `"typeof entity === "boolean""`;

exports[`GuardBaseTransformer constructs correct type guard for Type.Identifier 1`] = `"entity === TestEnum.ENUM_KEY_STRING"`;

exports[`GuardBaseTransformer constructs correct type guard for Type.InterfaceDeclaration 1`] = `"isTestInterface(entity)"`;

exports[`GuardBaseTransformer constructs correct type guard for Type.NumberKeyword 1`] = `"typeof entity === "number""`;

exports[`GuardBaseTransformer constructs correct type guard for Type.StringKeyword 1`] = `"typeof entity === "string""`;

exports[`GuardBaseTransformer constructs correct type guard for Type.TupleType 1`] = `"Array.isArray(entity) && Array.isArray(entity) && entity.every((item: any) => isTestEnum(item)) && entity.some(item => item === TestEnum.ENUM_KEY_STRING)"`;

exports[`GuardBaseTransformer constructs correct type guard for Type.TypeReference 1`] = `"(isGenericType || isTestInterfaceConstraintExtender)(entity)"`;

exports[`GuardBaseTransformer constructs correct type guard for Type.TypeReference 2`] = `"isTestTypeAlias(entity)"`;

exports[`GuardBaseTransformer constructs correct type guard for Type.TypeReference 3`] = `"isTestRelationship<TestNode>(entity, isTestNode)"`;

exports[`GuardBaseTransformer constructs correct type guard for Type.UnionType 1`] = `"(typeof entity === "string" || typeof entity === "number" || typeof entity === "boolean")"`;
